src/csharp/Microsoft.Spark/Utils/TypeConverter.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Spark.Utils
+{
+    /// <summary>
+    /// Converts one type to another.
+    /// </summary>
+    internal static class TypeConverter
+    {
+        /// <summary>
+        /// Convert obj to type <typeparamref name="T"/>.
+        /// </summary>
+        /// <typeparam name="T">Type to convert to</typeparam>
+        /// <param name="obj">The object to convert</param>
+        /// <returns></returns>
+        internal static T Convert<T>(object obj) => obj is T t ? t : (T)Convert(obj, typeof(T));
+
+        private static object Convert(object obj, Type toType)
+        {
+            if (obj == null)
+            {
+                return obj;
+            }
+
+            Type fromType = obj.GetType();
+            return fromType switch
+            {
+                _ when fromType == toType => obj,
+                _ when (fromType == typeof(ArrayList)) && toType.IsArray =>
+                    ConvertArrayList((ArrayList)obj, toType),
+                _ when (fromType == typeof(Hashtable)) && toType.IsGenericType &&
+                    (toType.GetGenericTypeDefinition() == typeof(Dictionary<,>)) =>
+                    ConvertHashtable((Hashtable)obj, toType),
+                _ => obj
```

Comment:
This seems to be an interesting case. Can we throw not implemented or no? Casting `int` to `long` can happen here?

Reply:
I think we don't need to throw anything.  It should error out at the caller if `obj` can't be cast to `(T)` or if it can't be implicitly converted when it's added into the Array/Dictionary.

Reply:
In that case, can you move casting here so that it's obvious? (and make this function to return `T` instead)

Reply:
We can't cast to `T` or return `T` here since `ConvertArrayList` and `ConvertHashtable` only has `Type` objects instead the actual types.  We can accomplish this if we used reflection.  Please see the [initial approach](https://github.com/dotnet/spark/pull/670/commits/c07a36bb6ed95160aef904829a008a6115807113#diff-45f40890dc2319e254fd60df2a87dc5bR12-R49), and we can change to that if you think it's better.

Reply:
Got it. This approach is fine.